File: tests/test_phase11_task2_governance.js
/**
 * PREDICTA — PHASE 11 TASK 2 HUMAN FEEDBACK GOVERNANCE & OFFLINE EVALUATION TEST SUITE (Node.js)
 * File: tests/test_phase11_task2_governance.js
 * 
 * Verifies Phase 11 Task 2 Requirements:
 * 1. Eligibility contract evaluation (ELIGIBLE_FOR_OFFLINE_REVIEW vs REJECTED_GOVERNANCE)
 * 2. Strict separation between operator feedback and ground truth (ground_truth_status === "NOT_ESTABLISHED")
 * 3. Conflict governance (conflicting operators yield REJECTED_GOVERNANCE, preserve all records)
 * 4. Offline evaluation candidate generation (evaluation_only: true, production_effect: false)
 * 5. Database-authoritative history reconstruction across cold starts / memory wipes
 * 6. Dataset leakage prevention & test set isolation
 * 7. Production protection (Model SHA 91bb59..., threshold 0.20, no retraining, no recalibration)
 */

process.env.NODE_ENV = 'test';

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const assert = require('assert');

const EXPECTED_MODEL_SHA = "91bb598ae91155674e40cb0a9f39d1e9bdeacd39875542db88b65e3668f29d98";
const EXPECTED_THRESHOLD = 0.20;

const PROD_MANIFEST_PATH = path.join(__dirname, '../ml/models/production/predicta_production_manifest.json');
const MODEL_JSON_PATH = path.join(__dirname, '../ml/models/production/predicta_xgboost_model.json');

const {
  HumanDispositionManagerJS,
  registerAuthoritativePrediction,
  DISPOSITION_CONTRACT_PATH
} = require('../src/governance/disposition');

console.log("=========================================================================");
console.log("🚀 PREDICTA — PHASE 11 TASK 2 GOVERNANCE & EVALUATION TEST SUITE (JS)");
console.log("=========================================================================\n");

let passed = 0;
let total = 0;

async function runTest(name, fn) {
  total++;
  try {
    await fn();
    console.log(`  ✓ [PASS] Test ${total.toString().padStart(2, '0')}: ${name}`);
    passed++;
  } catch (err) {
    console.error(`\n  ❌ [FAIL] Test ${total}: ${name}`);
    console.error(`     Reason: ${err.message}\n`);
    process.exit(1);
  }
}

async function runPhase11Task2JsTests() {
  const manager = new HumanDispositionManagerJS();

  // Test setup: Register authoritative predictions
  const sampleTraceId = "TRACE-P11T2-001";
  registerAuthoritativePrediction({
    trace_id: sampleTraceId,
    component_id: "COMP-P11T2-001",
    lot_id: "LOT-P11T2-001",
    prediction: "REJECT",
    probability: 0.85,
    anomaly_score: 0.92,
    prognostic_summary: "CRITICAL_DEGRADATION"
  });

  const conflictTraceId = "TRACE-P11T2-CONFLICT";
  registerAuthoritativePrediction({
    trace_id: conflictTraceId,
    component_id: "COMP-P11T2-CONF",
    lot_id: "LOT-P11T2-CONF",
    prediction: "REJECT",
    probability: 0.88,
    anomaly_score: 0.90
  });

  const unresolvedTraceId = "TRACE-P11T2-UNRESOLVED";
  registerAuthoritativePrediction({
    trace_id: unresolvedTraceId,
    component_id: "COMP-P11T2-UNRES",
    lot_id: "LOT-P11T2-UNRES",
    prediction: "ACCEPT",
    probability: 0.12
  });

  const benchmarkTraceId = "BENCHMARK_TRACE_001";
  registerAuthoritativePrediction({
    trace_id: benchmarkTraceId,
    component_id: "COMP-BENCH-01",
    lot_id: "LOT-BENCH-01",
    prediction: "REJECT",
    probability: 0.75
  });

  // -------------------------------------------------------------------------
  // TEST 1: Valid Candidate -> ELIGIBLE_FOR_OFFLINE_REVIEW
  // -------------------------------------------------------------------------
  await runTest("Valid complete record yields ELIGIBLE_FOR_OFFLINE_REVIEW", async () => {
    const disp = await manager.recordDispositionAsync({
      trace_id: sampleTraceId,
      disposition: "REJECT",
      reason_code: "FALSE_POSITIVE_SUSPECTED",
      operator_id: "OPERATOR_QUAL_01",
      comment: "Verified offline candidate candidate test"
    });

    await manager.updateFeedbackStatusAsync(sampleTraceId, disp.disposition_id, "CONFIRMED", "OPERATOR_QUAL_01", "Lab confirmation completed");

    const gov = await manager.evaluateDispositionGovernanceAsync(sampleTraceId);
    assert.strictEqual(gov.governance_classification, "ELIGIBLE_FOR_OFFLINE_REVIEW");
    assert.strictEqual(gov.rejection_reasons.length, 0);
    assert.ok(gov.evaluation_candidate);
    assert.strictEqual(gov.evaluation_candidate.trace_id, sampleTraceId);
    assert.strictEqual(gov.evaluation_candidate.ground_truth_status, "NOT_ESTABLISHED");
    assert.strictEqual(gov.evaluation_candidate.evaluation_only, true);
    assert.strictEqual(gov.evaluation_candidate.production_effect, false);
  });

  // -------------------------------------------------------------------------
  // TEST 2: Missing ML Provenance -> REJECTED_GOVERNANCE
  // -------------------------------------------------------------------------
  await runTest("Missing or corrupted ML model provenance yields REJECTED_GOVERNANCE", async () => {
    const badManager = new HumanDispositionManagerJS(
      DISPOSITION_CONTRACT_PATH,
      PROD_MANIFEST_PATH,
      MODEL_JSON_PATH
    );
    badManager.expectedModelSha = "0000000000000000000000000000000000000000000000000000000000000000";

    const gov = await badManager.evaluateDispositionGovernanceAsync(sampleTraceId);
    assert.strictEqual(gov.governance_classification, "REJECTED_GOVERNANCE");
    assert.ok(gov.rejection_reasons.includes("INVALID_MODEL_PROVENANCE"));
    assert.strictEqual(gov.evaluation_candidate, null);
  });

  // -------------------------------------------------------------------------
  // TEST 3: Missing Authoritative Prediction -> REJECTED_GOVERNANCE
  // -------------------------------------------------------------------------
  await runTest("Missing authoritative prediction record yields REJECTED_GOVERNANCE", async () => {
    const gov = await manager.evaluateDispositionGovernanceAsync("TRACE-NONEXISTENT-999");
    assert.strictEqual(gov.governance_classification, "REJECTED_GOVERNANCE");
    assert.ok(gov.rejection_reasons.includes("AUTHORITATIVE_ML_RECORD_NOT_FOUND"));
    assert.strictEqual(gov.evaluation_candidate, null);
  });

  // -------------------------------------------------------------------------
  // TEST 4: Client-Supplied ML Snapshot -> Rejected
  // -------------------------------------------------------------------------
  await runTest("Client-supplied ML snapshot injection attempt is rejected", async () => {
    let errOccurred = false;
    try {
      await manager.recordDispositionAsync({
        trace_id: sampleTraceId,
        disposition: "REJECT",
        reason_code: "FALSE_POSITIVE_SUSPECTED",
        probability: 0.10,
        ground_truth: "PASS"
      });
    } catch (err) {
      errOccurred = true;
      assert.ok(err.message.includes("CLIENT_CONTROLLED_ML_OUTPUT_PROHIBITED"));
    }
    assert.ok(errOccurred, "Expected error on client-supplied ML fields");
  });

  // -------------------------------------------------------------------------
  // TEST 5: Operator Feedback is Not Ground Truth
  // -------------------------------------------------------------------------
  await runTest("Operator CONFIRMED disposition does NOT set ground truth label", async () => {
    const gov = await manager.evaluateDispositionGovernanceAsync(sampleTraceId);
    assert.strictEqual(gov.evaluation_candidate.ground_truth_status, "NOT_ESTABLISHED");
    assert.notStrictEqual(gov.evaluation_candidate.ground_truth_status, "PASS");
    assert.notStrictEqual(gov.evaluation_candidate.ground_truth_status, "FAIL");
  });

  // -------------------------------------------------------------------------
  // TEST 6: Conflicting Operators -> REJECTED_GOVERNANCE
  // -------------------------------------------------------------------------
  await runTest("Conflicting operator dispositions yield REJECTED_GOVERNANCE with preserved ledger", async () => {
    const disp1 = await manager.recordDispositionAsync({
      trace_id: conflictTraceId,
      disposition: "ACCEPT",
      reason_code: "FALSE_POSITIVE_SUSPECTED",
      operator_id: "OPERATOR_01"
    });
    await manager.updateFeedbackStatusAsync(conflictTraceId, disp1.disposition_id, "CONFIRMED", "OPERATOR_01");

    const disp2 = await manager.recordDispositionAsync({
      trace_id: conflictTraceId,
      disposition: "REJECT",
      reason_code: "PROCESS_EXCEPTION",
      operator_id: "OPERATOR_02"
    });
    await manager.updateFeedbackStatusAsync(conflictTraceId, disp2.disposition_id, "CONFIRMED", "OPERATOR_02");

    const history = await manager.getDispositionAsync(conflictTraceId);
    assert.strictEqual(history.total_dispositions, 2);
    assert.strictEqual(history.has_conflict, true);

    const gov = await manager.evaluateDispositionGovernanceAsync(conflictTraceId);
    assert.strictEqual(gov.governance_classification, "REJECTED_GOVERNANCE");
    assert.ok(gov.rejection_reasons.some(r => r.includes("UNRESOLVED_GOVERNANCE_CONFLICT")));
    assert.strictEqual(gov.evaluation_candidate, null);
  });

  // -------------------------------------------------------------------------
  // TEST 7: Unresolved Lifecycle -> REJECTED_GOVERNANCE
  // -------------------------------------------------------------------------
  await runTest("Unresolved lifecycle status yields REJECTED_GOVERNANCE", async () => {
    const disp = await manager.recordDispositionAsync({
      trace_id: unresolvedTraceId,
      disposition: "HOLD",
      reason_code: "INSUFFICIENT_DATA",
      operator_id: "OPERATOR_03"
    });
    await manager.updateFeedbackStatusAsync(unresolvedTraceId, disp.disposition_id, "UNRESOLVED", "OPERATOR_03");

    const gov = await manager.evaluateDispositionGovernanceAsync(unresolvedTraceId);
    assert.strictEqual(gov.governance_classification, "REJECTED_GOVERNANCE");
    assert.ok(gov.rejection_reasons.some(r => r.includes("UNRESOLVED_LIFECYCLE_STATUS")));
  });

  // -------------------------------------------------------------------------
  // TEST 8: Missing Lifecycle History -> Fail Closed
  // -------------------------------------------------------------------------
  await runTest("Missing lifecycle history fails closed (REJECTED_GOVERNANCE)", async () => {
    const noEvtTraceId = "TRACE-P11T2-NOEVT";
    registerAuthoritativePrediction({
      trace_id: noEvtTraceId,
      component_id: "COMP-NOEVT",
      lot_id: "LOT-NOEVT",
      prediction: "REJECT",
      probability: 0.90
    });

    const disp = await manager.recordDispositionAsync({
      trace_id: noEvtTraceId,
      disposition: "REJECT",
      reason_code: "EQUIPMENT_ISSUE",
      operator_id: "OPERATOR_04"
    });

    // Clear lifecycle events map entry manually to simulate corrupt/missing lifecycle history
    const { _LIFECYCLE_EVENTS } = require('../src/governance/disposition');
    _LIFECYCLE_EVENTS.delete(disp.disposition_id);

    const gov = await manager.evaluateDispositionGovernanceAsync(noEvtTraceId);
    assert.strictEqual(gov.governance_classification, "REJECTED_GOVERNANCE");
    assert.ok(gov.rejection_reasons.some(r => r.includes("MISSING_LIFECYCLE_HISTORY")));
    assert.strictEqual(gov.evaluation_candidate, null);
  });

  // -------------------------------------------------------------------------
  // TEST A: Governed Evaluation without DB Client (require_durable_persistence=true) -> PERSISTENCE_ERROR
  // -------------------------------------------------------------------------
  await runTest("Governed evaluation without DB client and require_durable_persistence=true throws PERSISTENCE_ERROR", async () => {
    const noDbManager = new HumanDispositionManagerJS(
      DISPOSITION_CONTRACT_PATH,
      PROD_MANIFEST_PATH,
      MODEL_JSON_PATH,
      null
    );
    let thrown = false;
    try {
      await noDbManager.evaluateDispositionGovernanceAsync(sampleTraceId, { require_durable_persistence: true });
    } catch (err) {
      thrown = true;
      assert.ok(err.message.includes("PERSISTENCE_ERROR"), `Expected PERSISTENCE_ERROR, got: ${err.message}`);
    }
    assert.ok(thrown, "Expected evaluateDispositionGovernanceAsync to throw PERSISTENCE_ERROR when require_durable_persistence=true without DB client");
  });

  // -------------------------------------------------------------------------
  // TEST B / 9: Cold Start / Restart Reconstruction from Durable Storage
  // -------------------------------------------------------------------------
  await runTest("Restart reconstruction loads history from durable database client", async () => {
    const restartTraceId = "TRACE-RESTART-001";
    registerAuthoritativePrediction({
      trace_id: restartTraceId,
      component_id: "COMP-RESTART",
      lot_id: "LOT-RESTART",
      prediction: "REJECT",
      probability: 0.82
    });

    const mockSupabase = {
      from: (table) => {
        if (table === 'operator_dispositions') {
          return {
            select: () => ({
              eq: (col, val) => Promise.resolve({
                data: [{
                  disposition_id: "DISP-DURABLE-001",
                  trace_id: restartTraceId,
                  component_id: "COMP-RESTART",
                  lot_id: "LOT-RESTART",
                  operator_id: "OPERATOR_DURABLE",
                  disposition: "REJECT",
                  reason_code: "MANUAL_ENGINEERING_REVIEW",
                  created_at: new Date().toISOString(),
                  model_id_at_decision: "predicta_xgboost_model",
                  model_hash_at_decision: EXPECTED_MODEL_SHA,
                  original_ml_decision: "REJECT",
                  original_ml_probability: 0.82,
                  feedback_status: "CONFIRMED",
                  conflict: false
                }],
                error: null
              })
            })
          };
        }
        if (table === 'disposition_lifecycle_events') {
          return {
            select: () => ({
              eq: (col, val) => Promise.resolve({
                data: [{
                  event_id: "EVT-DURABLE-001",
                  disposition_id: "DISP-DURABLE-001",
                  trace_id: restartTraceId,
                  previous_status: "RECORDED_ONLY",
                  new_status: "CONFIRMED",
                  changed_by: "OPERATOR_DURABLE",
                  timestamp: new Date().toISOString()
                }],
                error: null
              })
            })
          };
        }
        return { select: () => ({ eq: () => Promise.resolve({ data: [], error: null }) }) };
      }
    };

    const durableManager = new HumanDispositionManagerJS(
      DISPOSITION_CONTRACT_PATH,
      PROD_MANIFEST_PATH,
      MODEL_JSON_PATH,
      mockSupabase
    );

    const gov = await durableManager.evaluateDispositionGovernanceAsync(restartTraceId);
    assert.strictEqual(gov.governance_classification, "ELIGIBLE_FOR_OFFLINE_REVIEW");
    assert.strictEqual(gov.evaluation_candidate.trace_id, restartTraceId);
    assert.strictEqual(gov.evaluation_candidate.lifecycle_status, "CONFIRMED");
  });

  // -------------------------------------------------------------------------
  // TEST D: isValidProbability rejects booleans
  // -------------------------------------------------------------------------
  await runTest("isValidProbability strictly rejects booleans (true/false)", async () => {
    const { isValidProbability } = require('../src/governance/disposition');
    assert.strictEqual(isValidProbability(true), false);
    assert.strictEqual(isValidProbability(false), false);
  });

  // -------------------------------------------------------------------------
  // TEST E: isValidProbability rejects NaN, Infinity, -Infinity
  // -------------------------------------------------------------------------
  await runTest("isValidProbability strictly rejects NaN, Infinity, -Infinity", async () => {
    const { isValidProbability } = require('../src/governance/disposition');
    assert.strictEqual(isValidProbability(NaN), false);
    assert.strictEqual(isValidProbability(Infinity), false);
    assert.strictEqual(isValidProbability(-Infinity), false);
  });

  // -------------------------------------------------------------------------
  // TEST F: isValidProbability bounds validation
  // -------------------------------------------------------------------------
  await runTest("isValidProbability bounds validation (-0.01, 1.01, string, null, 0.0, 1.0)", async () => {
    const { isValidProbability } = require('../src/governance/disposition');
    assert.strictEqual(isValidProbability(-0.01), false);
    assert.strictEqual(isValidProbability(1.01), false);
    assert.strictEqual(isValidProbability("0.5"), false);
    assert.strictEqual(isValidProbability(null), false);
    assert.strictEqual(isValidProbability(undefined), false);
    assert.strictEqual(isValidProbability(0.0), true);
    assert.strictEqual(isValidProbability(1.0), true);
    assert.strictEqual(isValidProbability(0.5), true);
  });

  // -------------------------------------------------------------------------
  // TEST G: extractOriginalMlDecision fallback order
  // -------------------------------------------------------------------------
  await runTest("extractOriginalMlDecision fallback order (prediction -> disposition -> decision)", async () => {
    const { extractOriginalMlDecision } = require('../src/governance/disposition');
    assert.strictEqual(extractOriginalMlDecision({ prediction: "REJECT", disposition: "ACCEPT", decision: "HOLD" }), "REJECT");
    assert.strictEqual(extractOriginalMlDecision({ disposition: "ACCEPT", decision: "HOLD" }), "ACCEPT");
    assert.strictEqual(extractOriginalMlDecision({ decision: "HOLD" }), "HOLD");
    assert.strictEqual(extractOriginalMlDecision({ prediction: null, disposition: "ACCEPT" }), "ACCEPT");
    assert.strictEqual(extractOriginalMlDecision({ prediction: undefined, disposition: null, decision: "HOLD" }), "HOLD");
    assert.strictEqual(extractOriginalMlDecision(null), null);
  });

  // -------------------------------------------------------------------------
  // TEST 10: Leakage Protection & Benchmark Isolation
  // -------------------------------------------------------------------------
  await runTest("Benchmark traces are protected against evaluation dataset leakage", async () => {
    const disp = await manager.recordDispositionAsync({
      trace_id: benchmarkTraceId,
      disposition: "REJECT",
      reason_code: "PROCESS_EXCEPTION",
      operator_id: "OPERATOR_05"
    });
    await manager.updateFeedbackStatusAsync(benchmarkTraceId, disp.disposition_id, "CONFIRMED", "OPERATOR_05");

    const gov = await manager.evaluateDispositionGovernanceAsync(benchmarkTraceId);
    assert.strictEqual(gov.governance_classification, "REJECTED_GOVERNANCE");
    assert.ok(gov.rejection_reasons.some(r => r.includes("TEST_SET_ISOLATION_PROTECTED")));
  });

  // -------------------------------------------------------------------------
  // TEST 11: Production Model Protection
  // -------------------------------------------------------------------------
  await runTest("Production model artifact SHA is unchanged", async () => {
    const content = fs.readFileSync(MODEL_JSON_PATH, 'utf-8').replace(/\r\n/g, '\n');
    const sha = crypto.createHash('sha256').update(content, 'utf8').digest('hex');
    assert.strictEqual(sha, EXPECTED_MODEL_SHA, "Production model SHA MUST remain untouched");
  });

  // -------------------------------------------------------------------------
  // TEST 12: Threshold Protection
  // -------------------------------------------------------------------------
  await runTest("Production operating threshold remains exactly 0.20", async () => {
    const manifest = JSON.parse(fs.readFileSync(PROD_MANIFEST_PATH, 'utf-8'));
    const threshold = manifest.authoritative_threshold || manifest.operating_threshold || manifest.threshold;
    assert.strictEqual(threshold, EXPECTED_THRESHOLD, "Threshold must remain 0.20");
  });

  // -------------------------------------------------------------------------
  // TEST 13 & 14: No Retraining & No Recalibration
  // -------------------------------------------------------------------------
  await runTest("No retraining or recalibration artifacts generated", async () => {
    const gov = await manager.evaluateDispositionGovernanceAsync(sampleTraceId);
    assert.strictEqual(gov.governance_guarantees.no_automatic_retraining, true);
    assert.strictEqual(gov.governance_guarantees.no_threshold_modification, true);
    assert.strictEqual(gov.governance_guarantees.no_conformal_recalibration, true);
  });

  console.log("\n=========================================================================");
  console.log(`✅ ALL ${passed}/${total} PHASE 11 TASK 2 JS GOVERNANCE TESTS PASSED SUCCESSFULLY!`);
  console.log("=========================================================================\n");
}

runPhase11Task2JsTests().catch(err => {
  console.error("Test execution failed:", err);
  process.exit(1);
});

File: tests/test_phase11_task2_governance.py
"""
PREDICTA — PHASE 11 TASK 2 HUMAN FEEDBACK GOVERNANCE & OFFLINE EVALUATION TEST SUITE (Python)
File: tests/test_phase11_task2_governance.py

Verifies Phase 11 Task 2 Requirements:
1. Eligibility contract evaluation (ELIGIBLE_FOR_OFFLINE_REVIEW vs REJECTED_GOVERNANCE)
2. Strict separation between operator feedback and ground truth (ground_truth_status === "NOT_ESTABLISHED")
3. Conflict governance (conflicting operators yield REJECTED_GOVERNANCE, preserve all records)
4. Offline evaluation candidate generation (evaluation_only: True, production_effect: False)
5. Database-authoritative history reconstruction across cold starts / memory wipes
6. Dataset leakage prevention & test set isolation
7. Production protection (Model SHA 91bb59..., threshold 0.20, no retraining, no recalibration)
8. JS / Python exact semantic parity
"""

import os
import sys
import json
import hashlib
import pytest
from unittest.mock import MagicMock

project_root = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
if project_root not in sys.path:
    sys.path.insert(0, project_root)

EXPECTED_MODEL_SHA = "91bb598ae91155674e40cb0a9f39d1e9bdeacd39875542db88b65e3668f29d98"
EXPECTED_THRESHOLD = 0.20

PROD_MANIFEST_PATH = os.path.join(project_root, "ml", "models", "production", "predicta_production_manifest.json")
MODEL_JSON_PATH = os.path.join(project_root, "ml", "models", "production", "predicta_xgboost_model.json")
DISPOSITION_CONTRACT_PATH = os.path.join(project_root, "ml", "governance", "disposition_contract.json")

from src.governance.disposition import (
    HumanDispositionManager,
    register_authoritative_prediction,
    _FEEDBACK_STORE,
    _LIFECYCLE_EVENTS
)


@pytest.fixture(autouse=True)
def setup_test_predictions():
    _FEEDBACK_STORE.clear()
    _LIFECYCLE_EVENTS.clear()

    register_authoritative_prediction({
        "trace_id": "TRACE-PY-T2-001",
        "component_id": "COMP-PY-001",
        "lot_id": "LOT-PY-001",
        "prediction": "REJECT",
        "probability": 0.85,
        "anomaly_score": 0.92,
        "prognostic_summary": "CRITICAL_DEGRADATION"
    })

    register_authoritative_prediction({
        "trace_id": "TRACE-PY-T2-CONFLICT",
        "component_id": "COMP-PY-CONF",
        "lot_id": "LOT-PY-CONF",
        "prediction": "REJECT",
        "probability": 0.88,
        "anomaly_score": 0.90
    })

    register_authoritative_prediction({
        "trace_id": "TRACE-PY-T2-UNRESOLVED",
        "component_id": "COMP-PY-UNRES",
        "lot_id": "LOT-PY-UNRES",
        "prediction": "ACCEPT",
        "probability": 0.12
    })

    register_authoritative_prediction({
        "trace_id": "BENCHMARK_PY_001",
        "component_id": "COMP-BENCH-PY-01",
        "lot_id": "LOT-BENCH-PY-01",
        "prediction": "REJECT",
        "probability": 0.75
    })


def test_01_valid_candidate_eligible():
    manager = HumanDispositionManager()
    disp = manager.record_disposition(
        trace_id="TRACE-PY-T2-001",
        disposition="REJECT",
        reason_code="FALSE_POSITIVE_SUSPECTED",
        operator_id="OPERATOR_PY_01",
        comment="Offline candidate verification"
    )
    manager.update_feedback_status(
        trace_id="TRACE-PY-T2-001",
        disposition_id=disp["disposition_id"],
        new_status="CONFIRMED",
        operator_id="OPERATOR_PY_01",
        comment="Lab confirmation completed"
    )

    gov = manager.evaluate_disposition_governance("TRACE-PY-T2-001")
    assert gov["governance_classification"] == "ELIGIBLE_FOR_OFFLINE_REVIEW"
    assert len(gov["rejection_reasons"]) == 0
    assert gov["evaluation_candidate"] is not None
    assert gov["evaluation_candidate"]["trace_id"] == "TRACE-PY-T2-001"
    assert gov["evaluation_candidate"]["ground_truth_status"] == "NOT_ESTABLISHED"
    assert gov["evaluation_candidate"]["evaluation_only"] is True
    assert gov["evaluation_candidate"]["production_effect"] is False


def test_02_missing_ml_provenance_rejected():
    bad_manager = HumanDispositionManager()
    bad_manager.expected_model_sha = "0000000000000000000000000000000000000000000000000000000000000000"

    gov = bad_manager.evaluate_disposition_governance("TRACE-PY-T2-001")
    assert gov["governance_classification"] == "REJECTED_GOVERNANCE"
    assert "INVALID_MODEL_PROVENANCE" in gov["rejection_reasons"]
    assert gov["evaluation_candidate"] is None


def test_03_missing_authoritative_prediction_rejected():
    manager = HumanDispositionManager()
    gov = manager.evaluate_disposition_governance("TRACE-PY-NONEXISTENT-999")
    assert gov["governance_classification"] == "REJECTED_GOVERNANCE"
    assert "AUTHORITATIVE_ML_RECORD_NOT_FOUND" in gov["rejection_reasons"]
    assert gov["evaluation_candidate"] is None


def test_04_client_supplied_ml_snapshot_rejected():
    manager = HumanDispositionManager()
    with pytest.raises(ValueError) as excinfo:
        manager.record_disposition(
            trace_id="TRACE-PY-T2-001",
            disposition="REJECT",
            reason_code="FALSE_POSITIVE_SUSPECTED",
            probability=0.10,
            ground_truth="PASS"
        )
    assert "CLIENT_CONTROLLED_ML_OUTPUT_PROHIBITED" in str(excinfo.value)


def test_05_operator_feedback_is_not_ground_truth():
    manager = HumanDispositionManager()
    disp = manager.record_disposition(
        trace_id="TRACE-PY-T2-001",
        disposition="REJECT",
        reason_code="FALSE_POSITIVE_SUSPECTED"
    )
    manager.update_feedback_status(
        trace_id="TRACE-PY-T2-001",
        disposition_id=disp["disposition_id"],
        new_status="CONFIRMED"
    )

    gov = manager.evaluate_disposition_governance("TRACE-PY-T2-001")
    assert gov["evaluation_candidate"]["ground_truth_status"] == "NOT_ESTABLISHED"
    assert gov["evaluation_candidate"]["ground_truth_status"] != "PASS"
    assert gov["evaluation_candidate"]["ground_truth_status"] != "FAIL"


def test_06_conflicting_operators_rejected():
    manager = HumanDispositionManager()
    disp1 = manager.record_disposition(
        trace_id="TRACE-PY-T2-CONFLICT",
        disposition="ACCEPT",
        reason_code="FALSE_POSITIVE_SUSPECTED",
        operator_id="OPERATOR_PY_01"
    )
    manager.update_feedback_status("TRACE-PY-T2-CONFLICT", disp1["disposition_id"], "CONFIRMED", "OPERATOR_PY_01")

    disp2 = manager.record_disposition(
        trace_id="TRACE-PY-T2-CONFLICT",
        disposition="REJECT",
        reason_code="PROCESS_EXCEPTION",
        operator_id="OPERATOR_PY_02"
    )
    manager.update_feedback_status("TRACE-PY-T2-CONFLICT", disp2["disposition_id"], "CONFIRMED", "OPERATOR_PY_02")

    history = manager.get_disposition("TRACE-PY-T2-CONFLICT")
    assert history["total_dispositions"] == 2
    assert history["has_conflict"] is True

    gov = manager.evaluate_disposition_governance("TRACE-PY-T2-CONFLICT")
    assert gov["governance_classification"] == "REJECTED_GOVERNANCE"
    assert any("UNRESOLVED_GOVERNANCE_CONFLICT" in r for r in gov["rejection_reasons"])
    assert gov["evaluation_candidate"] is None


def test_07_unresolved_lifecycle_rejected():
    manager = HumanDispositionManager()
    disp = manager.record_disposition(
        trace_id="TRACE-PY-T2-UNRESOLVED",
        disposition="HOLD",
        reason_code="INSUFFICIENT_DATA",
        operator_id="OPERATOR_PY_03"
    )
    manager.update_feedback_status("TRACE-PY-T2-UNRESOLVED", disp["disposition_id"], "UNRESOLVED", "OPERATOR_PY_03")

    gov = manager.evaluate_disposition_governance("TRACE-PY-T2-UNRESOLVED")
    assert gov["governance_classification"] == "REJECTED_GOVERNANCE"
    assert any("UNRESOLVED_LIFECYCLE_STATUS" in r for r in gov["rejection_reasons"])


def test_08_restart_reconstruction_from_durable_storage():
    restart_trace_id = "TRACE-RESTART-PY-001"
    register_authoritative_prediction({
        "trace_id": restart_trace_id,
        "component_id": "COMP-RESTART-PY",
        "lot_id": "LOT-RESTART-PY",
        "prediction": "REJECT",
        "probability": 0.82
    })

    mock_db = MagicMock()

    # Mock operator_dispositions query
    disp_query = MagicMock()
    disp_query.eq.return_value.execute.return_value.data = [{
        "disposition_id": "DISP-DURABLE-PY-001",
        "trace_id": restart_trace_id,
        "component_id": "COMP-RESTART-PY",
        "lot_id": "LOT-RESTART-PY",
        "operator_id": "OPERATOR_DURABLE",
        "disposition": "REJECT",
        "reason_code": "MANUAL_ENGINEERING_REVIEW",
        "created_at": "2026-09-22T00:00:00Z",
        "model_id_at_decision": "predicta_xgboost_model",
        "model_hash_at_decision": EXPECTED_MODEL_SHA,
        "original_ml_decision": "REJECT",
        "original_ml_probability": 0.82,
        "feedback_status": "CONFIRMED",
        "conflict": False
    }]
    disp_query.eq.return_value.execute.return_value.error = None

    # Mock disposition_lifecycle_events query
    evt_query = MagicMock()
    evt_query.eq.return_value.execute.return_value.data = [{
        "event_id": "EVT-DURABLE-PY-001",
        "disposition_id": "DISP-DURABLE-PY-001",
        "trace_id": restart_trace_id,
        "previous_status": "RECORDED_ONLY",
        "new_status": "CONFIRMED",
        "changed_by": "OPERATOR_DURABLE",
        "timestamp": "2026-09-22T00:00:00Z"
    }]
    evt_query.eq.return_value.execute.return_value.error = None

    def table_router(table_name):
        if table_name == "operator_dispositions":
            mock_t = MagicMock()
            mock_t.select.return_value = disp_query
            return mock_t
        if table_name == "disposition_lifecycle_events":
            mock_t = MagicMock()
            mock_t.select.return_value = evt_query
            return mock_t
        return MagicMock()

    mock_db.table.side_effect = table_router

    durable_manager = HumanDispositionManager(db_client=mock_db)

    gov = durable_manager.evaluate_disposition_governance(restart_trace_id)
    assert gov["governance_classification"] == "ELIGIBLE_FOR_OFFLINE_REVIEW"
    assert gov["evaluation_candidate"]["trace_id"] == restart_trace_id
    assert gov["evaluation_candidate"]["lifecycle_status"] == "CONFIRMED"


def test_09_leakage_protection_benchmark_isolation():
    manager = HumanDispositionManager()
    disp = manager.record_disposition(
        trace_id="BENCHMARK_PY_001",
        disposition="REJECT",
        reason_code="PROCESS_EXCEPTION",
        operator_id="OPERATOR_PY_05"
    )
    manager.update_feedback_status("BENCHMARK_PY_001", disp["disposition_id"], "CONFIRMED", "OPERATOR_PY_05")

    gov = manager.evaluate_disposition_governance("BENCHMARK_PY_001")
    assert gov["governance_classification"] == "REJECTED_GOVERNANCE"
    assert any("TEST_SET_ISOLATION_PROTECTED" in r for r in gov["rejection_reasons"])


def test_10_production_model_protection():
    with open(MODEL_JSON_PATH, "r", encoding="utf-8") as f:
        content = f.read().replace("\r\n", "\n")
    sha = hashlib.sha256(content.encode("utf-8")).hexdigest()
    assert sha == EXPECTED_MODEL_SHA, "Production model SHA must remain untouched"


def test_11_threshold_protection():
    with open(PROD_MANIFEST_PATH, "r", encoding="utf-8") as f:
        manifest = json.load(f)
    threshold = manifest.get("authoritative_threshold") or manifest.get("operating_threshold") or manifest.get("threshold")
    assert threshold == EXPECTED_THRESHOLD, "Production threshold must remain 0.20"


def test_12_no_retraining_no_recalibration():
    manager = HumanDispositionManager()
    gov = manager.evaluate_disposition_governance("TRACE-PY-T2-001")
    guarantees = gov["governance_guarantees"]
    assert guarantees["no_automatic_retraining"] is True
    assert guarantees["no_threshold_modification"] is True
    assert guarantees["no_conformal_recalibration"] is True
